### PR TITLE
Pre-filter example function allows filtering against all CloudEvent data

### PR DESCRIPTION
**Pull request by dsbibby**
* Refactored the pre-filter function to enable filtering against all data fields and make it leaner and faster
* Can now specify if all defined filters must match
* Handles traversing in to dicts/lists in the event data
* Added functionality to use 'n' in numeric indices of lists
* Added faas stack tracing support
* Unit tests significantly improved
* Documentation updates and improved logging

## Summary

These changes allow the pre-filter function to be used to apply filtering against any data in the received CloudEvent data, including traversing in to dicts and lists in the event data.

The original ability to filter against vcenter object paths is retained and doesn't change.

## Pull Request Checklist
🚨 Please review the [guidelines for contributing](https://vmweventbroker.io/community) to this repository.

- [x] Please ensure that you are making a pull request against the **Development** branch
- [ ] Please use the `WIP` keyword in the title of your PR if you are not ready for review
- [ ] Please ensure that you have opened a Github Issue if you are resolving/fixing a problem
- [x] Please ensure that you have [signed](https://help.github.com/en/github/authenticating-to-github/signing-commits) all commits and that you have [squashed](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) all relevant commits related to your change
- [ ] Please make sure that you have tested your change locally by successfully [building and deploying the VMware Event Broker Appliance](https://vmweventbroker.io/kb/contribute-appliance) and/or [building and deploying VMware Event Router](https://vmweventbroker.io/kb/contribute-eventrouter)
- [x] Please include any relevant screenshots and/or output as part of your testing
- [x] Please include any documentation updates that is applicable for your changes

## Change Type

What types of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the type of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation changes
- [ ] Other (please describe):

## Resolved Issues

List of Issues closed or resolved by this PR. Add multiple `Closes` keyword followed by the issue number (e.g. Closes #ISSUE-NUMBER)

* N/A

## Testing Verification

* Full set of unit tests now included and ran successfully:
```$ python3 pre-filter/unit_tests.py 2> /dev/null
TEST ID 1 PASSED: 'Invalid JSON > JSONDecodeError: Expecting value: line 1 column 1 (char 0)' - 500
TEST ID 2 PASSED: 'Invalid JSON > JSONDecodeError: Extra data: line 1 column 7 (char 6)' - 500
TEST ID 3 PASSED: 'Invalid JSON, required key not found > KeyError: 'data'' - 500
TEST ID 4 PASSED: 'Invalid JSON, data not iterable > TypeError: string indices must be integers' - 500
TEST ID 5 PASSED: 'some filters not matched' - 400
TEST ID 6 PASSED: 'call_function must be specified' - 400
TEST ID 7 PASSED: 'Index out of range' - 400
TEST ID 8 PASSED: 'call_function must be specified' - 400
TEST ID 9 PASSED: 'Filter "gree$" does not match any value' - 200
TEST ID 10 PASSED: 'call_function must be specified' - 400
TEST ID 11 PASSED: 'Filter "/.*/hos/" does not match "/DC/host/Lab/esxi03.lab"' - 200
TEST ID 12 PASSED: 'call_function must be specified' - 400
TEST ID 13 PASSED: 'call_function must be specified' - 400
TEST ID 14 PASSED: 'call_function must be specified' - 400
ALL TESTS PASSED!```

## Additional Information

* Any other details you wish to include or mention

If you have any questions/comments, feel free to reach out to team on Slack [#vcenter-event-broker-appliance](https://vmwarecode.slack.com/archives/CQLT9B5AA)

Thank you from the VEBA Team! 🥳

### CODE REVIEW

**Comment by dsbibby** [comment]
@lamw @embano1 Is it possible to remove the vmwclabot comment above that references my corporate email address? I rather unfortunately had the commit author set wrong!

**Comment by embano1** [comment]
you can amend the commit and change author, sign, etc.

**Comment by dsbibby** [comment]
@embano1 Yes, I've corrected the commit already - just wondered if it were possible to remove the auto comment made by the vmwclabot above - I can't amend that sadly.

**Comment by embano1** [comment]
don't worry about the comment, the label is already removed thus not blocking

File: examples/python/pre-filter/pre-filter/handler.py [hunk 1]
```diff
@@ -190,37 +330,56 @@ def handle(req):
...
+        if 'faasstack' in cevent:
+            cevent['faasstack'].append(faasFunctionName())
+        else:
+            cevent['faasstack'] = [faasFunctionName()]
+
+        # Send a post request to the chained function with the original event data
         try:
-            res = requests.post(f'http://gateway.openfaas:8080/function/{func}', req)
+            res = requests.post(f'http://gateway.openfaas:8080/function/{func}', json.dumps(cevent))
             return res.text, res.status_code
```
Comment (embano1) [inline examples/python/pre-filter/pre-filter/handler.py:383]: This works nicely with the retries we're adding in #241 so you get automatic retries for free here.

**Review by embano1 — COMMENTED** [review]
Really nice work, just some minor comments awaiting your feedback.

File: examples/python/pre-filter/README.md [hunk 2]
```diff
@@ -70,8 +71,9 @@ functions:
     environment:
       write_debug: true
       read_debug: true
-      insecure_ssl: true
-      call_function: veba-echo
+      insecure_ssl: true # set to true to disable validation of vcenter ssl certificate
```
[the remark this answers is not in the document]
Comment (dsbibby) [reply examples/python/pre-filter/README.md:74]: Certainly, will update this.

File: examples/python/pre-filter/pre-filter/handler.py [hunk 3]
```diff
@@ -150,25 +172,143 @@ def filterVcObject(obj, filter):
...
+        debug(f'{bgc.FAIL}Invalid regex pattern specified - {err.msg} at pos {err.pos}{bgc.ENDC}')
+        return f'Invalid regex pattern specified - {err.msg} at pos {err.pos}', 500
+    except TypeError:
+        debug(f'{bgc.FAIL}Cannot apply filter against non-string object. Type is: {typeName(value)}{bgc.ENDC}')
+        return f'Cannot apply filter against non-string object. Type is: {typeName(value)}', 200
+
+
+def filterItem(key, filter, data):
+    """
+    Try to filter against the passed value
```
[the remark this answers is not in the document]
Comment (dsbibby) [reply examples/python/pre-filter/pre-filter/handler.py:207]: Yep - I'll update all the docstrings to proper [Google Style](https://google.github.io/styleguide/pyguide.html) as it looks like I should be using `Args:` not `Parameters:` too.

On this, do you prefer the docstring to detail param types, or for the function definition to include type annotations, i.e.:
```python
def filterItem(key: str, filter: str, data: object):
    """
    Try to filter against the passed value
```

[the remark this answers is not in the document]
Comment (dsbibby) [reply examples/python/pre-filter/pre-filter/handler.py]: I've committed updates to just the docstrings for now. Let me know if you'd rather have type annotations too.

**Review by embano1 — COMMENTED** [review]
Excellent! Thx a ton! (and sorry for the delay)
LGTM. Please squash and I'll approve.

**Comment by dsbibby** [comment]
Thanks @embano1 - squashed and force pushed.

**Review by embano1 — APPROVED** [review]
LGTM

**Comment by embano1** [comment]
Thx!

Minor ask: next time please follow git commit naming/text best practices.

**Pull request merged**